soc/arm/microchip_mec/mec172x/soc.c

```diff
@@ -0,0 +1,417 @@
+			break;
+		case CLK32K_SRC_CRYSTAL:
+			vbr->CLK32_SRC = vbcs | MCHP_VBATR_CS_PCS_VTR_VBAT_XTAL;
+			break;
+		case CLK32K_SRC_PIN:
+			if (flags & CLK32K_FLAG_PIN_FB_CRYSTAL) {
+				vbr->CLK32_SRC =
+					vbcs | MCHP_VBATR_CS_PCS_VTR_PIN_SO;
+			} else {
+				vbr->CLK32_SRC =
+					vbcs | MCHP_VBATR_CS_PCS_VTR_PIN_XTAL;
+			}
+			break;
+		default: /* do not touch HW */
+			break;
+		}
+	}
+}
+
+static int pll_wait_lock(uint32_t wait_cnt)
+{
+	struct pcr_regs *regs = PCR_XEC_REG_BASE;
+
+	while (!(regs->OSC_ID & MCHP_PCR_OSC_ID_PLL_LOCK)) {
+		if (wait_cnt == 0) {
+			return -ETIMEDOUT;
+		}
+		--wait_cnt;
+	}
+
+	return 0;
+}
+
+/*
+ * MEC172x has two 32 KHz clock domains
+ *   PLL domain: 32 KHz clock input for PLL to produce 96 MHz and 48 MHz clocks
+ *   Peripheral domain: 32 KHz clock for subset of peripherals.
+ * Each domain 32 KHz clock input can be from one of the following sources:
+ *   Internal Silicon oscillator: +/- 2%
+ *   External Crystal connected as parallel or single ended
+ *   External 32KHZ_PIN 50% duty cycle waveform with fall back to either
+ *     Silicon OSC or crystal when 32KHZ_PIN signal goes away or VTR power rail
+ *     goes off.
+ * At chip reset the PLL is held in reset and the +/- 50% ring oscillator is
+ * the main clock.
+ * If no VBAT reset occurs the VBAT 32 KHz soure register maintains its state.
+ */
+static int soc_clk32_init(enum clk32k_src pll_clk_src,
+			  enum clk32k_src periph_clk_src,
+			  uint32_t flags)
+{
+	if (!is_sil_osc_enabled()) {
+		enable_sil_osc(CLK32K_SIL_OSC_DELAY);
+	}
+
+	/* Default to 32KHz Silicon OSC for PLL and peripherals */
+	connect_32k_source(CLK32K_SRC_SIL_OSC,  CLK32K_DEST_PLL, 0);
+	connect_32k_source(CLK32K_SRC_SIL_OSC, CLK32K_DEST_PERIPH, 0);
+
+	int rc = pll_wait_lock(CLK32K_PLL_LOCK_WAIT);
+
+	if (rc) {
+		return rc;
+	}
+
+	/*
+	 * We only allow Silicon OSC or Crystal as PLL source.
+	 * Any of the three source is allowed for Peripheral clock
+	 */
+	if ((pll_clk_src == CLK32K_SRC_CRYSTAL) ||
+	    (periph_clk_src == CLK32K_SRC_CRYSTAL)) {
+		enable_32k_crystal(flags);
+		rc = check_32k_crystal();
+		if (rc) {
+			return rc;
+		}
+		if (pll_clk_src == CLK32K_SRC_CRYSTAL) {
+			connect_32k_source(CLK32K_SRC_CRYSTAL, CLK32K_DEST_PLL,
+					   flags);
+		}
+		if (periph_clk_src == CLK32K_SRC_CRYSTAL) {
+			connect_32k_source(CLK32K_SRC_CRYSTAL,
+					   CLK32K_DEST_PERIPH,
+					   flags);
+		}
+		rc = pll_wait_lock(CLK32K_PLL_LOCK_WAIT);
+	} else if (periph_clk_src == CLK32K_SRC_PIN) {
+		rc = enable_32k_pin(CLK32K_PIN_WAIT);
+		if (rc) {
+			return rc;
+		}
+		connect_32k_source(CLK32K_SRC_PIN, CLK32K_DEST_PERIPH, flags);
+	}
+
+	return rc;
+}
+
+uint32_t soc_get_core_clock(void)
+{
+	struct pcr_regs *regs = PCR_XEC_REG_BASE;
+
+	return (uint32_t)(MCHP_EC_CLOCK_INPUT_HZ) / regs->PROC_CLK_CTRL;
+}
+
+/* Implementation based upon MEC172x Errata */
+void soc_set_core_clock_div(uint8_t clkdiv)
+{
+	struct pcr_regs *regs = PCR_XEC_REG_BASE;
+
+	arch_nop();
```

Comment:
what are the nops for? please add a comment

Reply:
Hi @sjg20 Added comments referring to the MEC172x errata requiring the specific code sequence when the PCR clock divider is changed.